mycroft/skills/event_scheduler.py

```diff
@@ -250,3 +251,163 @@ def shutdown(self):
+            name = self.skill.name + handler.__name__
+        unique_name = self._unique_name(name)
+        if repeat:
+            self.scheduled_repeats.append(name)  # store "friendly name"
+
+        data = data or {}
+        self.skill.add_event(unique_name, handler, once=not repeat)
+        event_data = {}
+        event_data['time'] = time.mktime(when.timetuple())
+        event_data['event'] = unique_name
+        event_data['repeat'] = repeat
+        event_data['data'] = data
+        self.skill.bus.emit(Message('mycroft.scheduler.schedule_event',
+                                    data=event_data))
+
+    def schedule_event(self, handler, when, data=None, name=None):
+        """Schedule a single-shot event.
+
+        Arguments:
+            handler:               method to be called
+            when (datetime/int/float):   datetime (in system timezone) or
+                                   number of seconds in the future when the
+                                   handler should be called
+            data (dict, optional): data to send when the handler is called
+            name (str, optional):  reference name
+                                   NOTE: This will not warn or replace a
+                                   previously scheduled event of the same
+                                   name.
+        """
+        data = data or {}
+        if isinstance(when, (int, float)):
+            when = datetime.now() + timedelta(seconds=when)
+        self._schedule_event(handler, when, data, name)
+
+    def schedule_repeating_event(self, handler, when, frequency,
+                                 data=None, name=None):
+        """Schedule a repeating event.
+
+        Arguments:
+            handler:                method to be called
+            when (datetime):        time (in system timezone) for first
+                                    calling the handler, or None to
+                                    initially trigger <frequency> seconds
+                                    from now
+            frequency (float/int):  time in seconds between calls
+            data (dict, optional):  data to send when the handler is called
+            name (str, optional):   reference name, must be unique
+        """
+        # Do not schedule if this event is already scheduled by the skill
+        if name not in self.scheduled_repeats:
+            data = data or {}
+            if not when:
+                when = datetime.now() + timedelta(seconds=frequency)
+            self._schedule_event(handler, when, data, name, frequency)
+        else:
+            LOG.debug('The event is already scheduled, cancel previous '
+                      'event if this scheduling should replace the last.')
+
+    def update_scheduled_event(self, name, data=None):
+        """Change data of event.
+
+        Arguments:
+            name (str): reference name of event (from original scheduling)
+        """
+        data = data or {}
```

Comment:
Not sure I understand your comment of passing by reference. Reassigning a label does not change the origin.

I think it's good practice to explicitly show how a default value is applied in a method. Since dict is a mutable type we shouldn't use `data={}` in the argument line. Instead it's defaulted to `None` which is immutable and safe. Then as the first step in the method the default is applied.

Reply:
Tabling this for now.